Section 2: Simple chat npcs (#11)
RaiderSoft/semantic-godot: docs/first_semantic_game/02.simple_chat_npcs.rst

Simple Chat NPCs
=================

This section aims to guide you through the process of creating NPCs that 
can hold interactive conversations.  It assumes that you have completed the 
project setup and have configured the OPENAI_API_KEY user secret.

The NPC Sprite
--------------

Here we will create a basic sprite that we can attach chat functionality to.

1. In the ``MainGame`` scene that we created in :doc:`01.project_setup`
   insert an ``Area2D`` node and rename it to  ``NPC``.
2. Add a ``CollisionShape2D`` rectangle to the NPC. This will be useful 

>    later and removes the yellow warning in the workspace.

Reviewer: I think it would be fine to make this a `Node2D` at this point and have the `ChatEntity` class inherit from `Node2D` for now. We don't need the functionality of the `Area2D` right now, so not worrying about that would make this section a little simpler and clearer. It should be easy enough to change the `ChatEntity` script later so that it inherits from `Area2D` when we need that functionality.

Reviewer: I just saw [<https://github.com/RaiderSoft/semantic-godot/issues/5#issuecomment-1710460506>](url) about using the NPC scene as a starting point for the Gralk scene. I think this is an option, but it might add some complexity to this first demo part.

If we did this, we'd want to make the root `NPC` node a `CharacterBody2D`, which is what it is for the Gralk scene later on. Also, we might want to attach `ChatAI` as a child node of the `NPC` node (not as a script for the `NPC` node), since that is how we do it later on. To minimize complexity, the `ChatEntity`/`ChatAI` nodes could still just inherit from `Node2D` for this section.

Author: Changed to Node2D

3. With the ``NPC`` node selected, add a ``Sprite2D`` node. This 
   makes a visible sprite for our character.
4. To set the image for our NPC sprite, copy the troll portrait into 
   the project. Then locate and choose the image as an ``ImageTexture`` 
   in the properties panel for the ``Sprite2D`` node.
5. Drag the NPC image near the center of the scene. Build and run the project to
   see the river troll NPC token.

Meet Gralk:

.. image:: gralk.png
   :alt: The Gralk NPC sprite
   :width: 150px

>    :align: center

Author: Agreed but skipping in this PR.

The ChatEntity class
--------------------

Here we a define a base class called ``ChatEntity`` for all chat-enabled  entities in the game. 
This type encapsulates basic properties like a name and a description, and, later, 
it handles behaviors such as detecting other nearby chat entities. Subsequently, 
we will extend this node type to create ``ChatAI`` for GPT-4 driven NPCs and 
``ChatPlayer`` for player characters.

Right-click the filesystem dock and select "New Script...". **Make sure the language 
is set to C#** and it inherits from ``Area2D``.

Save this script as ``ChatEntity.cs`` in the project's root directory.

> 

Reviewer: Should we insert the instructions about adding the `[GlobalClass]` attribute here?

Reviewer: ```suggestion

To register this script with the Godot editor, add the ``[GlobalClass]`` attribute right above the class declaration.

.. code-block:: csharp

    [GlobalClass]
    public partial class ChatEntity : Area2D
    {
        ...
    }

For more information about Godot's global classes, see https://docs.godotengine.org/en/stable/tutorials/scripting/c_sharp/c_sharp_global_classes.html

```

Exported Properties
^^^^^^^^^^^^^^^^^^^

> 

Reviewer: Should we insert the instructions about adding the `[GlobalClass]` attribute here?

Next, we will add methods for handling messages: ``SendMsg`` and ``ReceiveMsg``.

The ``SendMsg`` method is called when the entity wants to send a message to another
entity. It prints the message to the console and then emits a ``MsgSent`` signal
with the message as an argument. 

``ReceiveMsg`` is called when the entity receives a message from another entity.
We expect this method to be overridden by a child class because the behavior of
receiving a message will be different for different types of entities.

.. code-block:: csharp

    [Signal]
    public delegate void MsgSentEventHandler(string msg);

    public virtual void SendMsg(string msg)
    {
        GD.Print(ChatName + ": " + msg);
        CallDeferred("emit_signal", "MsgSent", msg);
    }

    public virtual void ReceiveMsg(string msg)
    {
        // This function will be overridden later by a child class
    }

Finally, we need to add a field to store a reference to our current conversation
partner. This will be used later to send messages to the other entity.
This ``_inConvoWith`` field is declared as ``internal`` so that it can be accessed

> by the main game class.

Reviewer: I accidentally skipped over this the first time through. Maybe add a code block just for the `_inConvoWith` field declaration so that this part is harder to miss?

Reviewer: ```suggestion
Finally, we need to add a field to store a reference to our current conversation
partner. This will be used later to send messages to the other entity.
This ``_inConvoWith`` field is declared as ``internal`` so that it can be accessed
by the main game class.

.. code-block:: csharp

    internal ChatEntity _inConvoWith;

```

Now the completed ``ChatEntity`` class should look like this:

.. code-block:: csharp

    using Godot;
    using System;

>     public partial class ChatEntity : Area2D

Reviewer: I had an issue where the `ChatName` and `ChatDescr` properties were not exported to the inspector. I changed this `ChatEntity` class to be a `[GlobalClass]`, rebuilt the project, and that fixed it. So I think we'll need to make sure the tutorial-user knows to add the `[GlobalClass]` attribute for `ChatEntity`.

Reviewer: ```suggestion

    [GlobalClass]
    public partial class ChatEntity : Area2D
```

Reviewer: Also made a suggestion [here](https://github.com/RaiderSoft/semantic-godot/pull/11#discussion_r1320190964) about where we can include instructions about adding the `[GlobalClass]` attribute.

            ChatRequestSettings settings = new();
            string reply = await _chatGPT.GenerateMessageAsync(_chat, settings);

            // Send the reply to the other entity via the MsgSent signal
            SendMsg(reply);

            // Add the reply to the chat history.
            _chat.AddAssistantMessage(reply);
        }
        catch (Exception ex)
        {
            string errMsg = $"Could not get reply from {ChatName}: {ex.Message}";
            GD.PrintErr(errMsg);
        }
    }

Two Arguing NPCs
----------------

In this section, you will learn how to set up NPCs named "East Troll" and 
"West Troll" to automatically argue with each other using Godot's properties
panel. These trolls are configured to have a debate about which side of the 
river is better.

Creating the East and West Troll Nodes
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

1. **Create an NPC Scene**: Drag the ``NPC`` node from the main scene into 
   the filesystem dock to create a new scene. This will convert the node
   in the main scene to an instance of the new scene.
2. **Duplicate the NPC**: Create a second troll by right-clicking the 

>    ``NPC.tscn`` scene in the filesystem dock and selecting "Duplicate". 

Reviewer: In the `MainGame` scene, we can create two instances of the same `NPC` scene and change their properties (node name, `ChatName`, and `ChatDescr`). So I don't think we'll have to duplicate `NPC.tscn` here.

Reviewer: ```suggestion
2. **Second NPC**: Create a second troll by dragging the 
   ``NPC.tscn`` scene into the scene tree. This should add a
   second instance of the ``NPC`` node to the main scene. 
```

> 3. **Rename Nodes**: Name the nodes ``West Troll`` the nodes ``East Troll``. Place the tokens on the left and right sides of the scene, respectively.

Reviewer: I think the naming convention for nodes in Godot is PascalCase, so maybe change these to `WestTroll` and `EastTroll`. Would also have to change the MainGame.cs script, as noted.

Reviewer: ```suggestion
3. **Rename Nodes**: Name the nodes ``WestTroll`` the nodes ``EastTroll``. Place the tokens on the left and right sides of the scene, respectively.
```

Troll Personalities
^^^^^^^^^^^^^^^^^^^

1. **Edit East Troll ChatDescr**: Find the ``ChatDescr`` 
   in the properties panel. Here, input the description: "I engage in a 
   passionate debate with the West Troll over why my side of the river is 
   superior."

2. **Edit West Troll ChatDescr**: Similar to the East Troll, enter a description that 
   states, "I argue with the East Troll about why my side of the river is 
   far better."

3. **Save and Run**: Save the scene and run your game to ensure that the 
   troll NPCs are in place and ready to argue.

Modify ``MainGame`` to connect the two trolls together.

.. code-block:: csharp

   using Godot;
   using Microsoft.SemanticKernel;

   public partial class MainGame : Node
   {
       public override void _Ready()
       {
           GD.Print("Semantic Kernel is ready!");

           // Get the troll ChatEntities
           ChatEntity eastTroll = GetNode<ChatEntity>("East Troll");

>            ChatEntity westTroll = GetNode<ChatEntity>("West Troll");

Reviewer: If changing the names of the nodes, will also have to change these lines to get nodes `"EastTroll"` and `"WestTroll"`.

Reviewer: ```suggestion
           ChatEntity eastTroll = GetNode<ChatEntity>("EastTroll");
           ChatEntity westTroll = GetNode<ChatEntity>("WestTroll");
```
